internal/pkg/api/plugin.go

```diff
@@ -0,0 +1,74 @@
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package api
+
+import "io"
+
+// The following comments is to illustrate the relationship between different plugin interface in api package.
+//
+//
+//                                                 Processors
+//                                   -----------------------------------------
+//  ---------        ---------        -----------                 -----------
+// | Gatherer | ==> |  Queue   | ==> | Processor | ==>  ...  ==> | Processor |
+// | (Parser) |     | Mem/File |      -----------                 -----------
+//  ----------       ---------            ||                          ||
+//                                        \/	                      \/
+//                                    ---------------------------------------
+//                                   |             OutputEventContext        |
+//                                    ---------------------------------------
+//                                                                    ||
+//                                                                    \/
+//                                    -------------------       -------------
+//                                   | BatchOutputEvents | <== | BatchBuffer |
+//                                    -------------------       -------------
+//                                             ||
+//                                             \/
+//                                    -------------------
+//                                   |     Forwarder     | ==> Kakfa/OAP
+//                                    -------------------
+// 1. The Gatherer plugin would fetch or receive the input data.
+// 2. The Parser plugin would parse the input data to InputEvent.
+//    If the event needs output, please tag it by the IsOutput
+//    method.
+// 3. The Queue plugin would store the InputEvent. But different
+//    Queue would use different ways to store data, such as store
+//    bytes by serialization or keep original.
+// 4. The Processor plugin would pull the event from the Queue and
+//    process the event to create a new event. Next, the event is
+//    passed to the next processor to do the same things until the
```

Comment:
What do you mean by the next processor? Do you plan to support multiple consumers of gatherer queue?